File: javascript/dsa/data-structures/binary-tree/flatten-binary-tree.js
/** Definition for a binary tree node */
// class TreeNode { val, left, right }

var flatten = function(root) { // @viz: function_start
    let prev = null;
    
    const traverse = (node) => {
        if (!node) return; // @viz: base_case
        
        traverse(node.right); // @viz: recurse_right
        traverse(node.left); // @viz: recurse_left
        
        node.right = prev; // @viz: update_right
        node.left = null; // @viz: update_left
        prev = node; // @viz: update_prev
    };
    
    traverse(root); // @viz: function_start
};

File: go/dsa/data-structures/binary-tree/flatten-binary-tree.go
/** Definition for a binary tree node */
// type TreeNode struct { Val int; Left, Right *TreeNode }

var prev *TreeNode

func flatten(root *TreeNode) { // @viz: function_start
    prev = nil
    var traverse func(*TreeNode)
    traverse = func(node *TreeNode) {
        if node == nil { return } // @viz: base_case
        
        traverse(node.Right) // @viz: recurse_right
        traverse(node.Left) // @viz: recurse_left
        
        node.Right = prev // @viz: update_right
        node.Left = nil // @viz: update_left
        prev = node // @viz: update_prev
    }
    traverse(root) // @viz: function_start
}